File: src/main/java/com/liamfarrell/android/snapbattle/activity/startup/ChooseUsernameStartupFragment.java
package com.liamfarrell.android.snapbattle.activity.startup;



import android.app.Activity;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.mobileconnectors.lambdainvoker.LambdaFunctionException;
import com.amazonaws.mobileconnectors.lambdainvoker.LambdaInvokerFactory;
import com.amazonaws.regions.Regions;
import com.liamfarrell.android.snapbattle.model.AsyncTaskResult;
import com.liamfarrell.android.snapbattle.model.aws_lambda_function_deserialization.aws_lambda_functions.LambdaFunctionsInterface;
import com.liamfarrell.android.snapbattle.R;
import com.liamfarrell.android.snapbattle.activity.FacebookLoginFragment;
import com.liamfarrell.android.snapbattle.model.aws_lambda_function_deserialization.aws_lambda_functions.response.UpdateUsernameResponse;
import com.liamfarrell.android.snapbattle.model.lambda_function_request_objects.UpdateUsernameRequest;
import com.liamfarrell.android.snapbattle.util.HandleLambdaError;

import java.lang.ref.WeakReference;

/**
 * Created by Liam on 1/01/2018.
 */

public class ChooseUsernameStartupFragment extends Fragment
{
    public final static String  usernamameNotValidErorrCode = "USERNAME_NOT_VALID";
    public final static String  usernameTooLongErrorCode = "USERNAME_TOO_LONG";
    public final static String  usernameUsernameAlreadyExistsErrorCode = "USERNAME_ALREADY_EXISTS";

    private EditText mUsernameEditText;
    private String mDefaultUsername;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        mDefaultUsername = ((StartupActivity)getActivity()).getUsername();
        View v = inflater.inflate(R.layout.fragment_choose_username_startup, container, false);
        mUsernameEditText = v.findViewById(R.id.usernameEditText);
        //username will be premade on server, user can update it if they want
        mUsernameEditText.setText(mDefaultUsername);


        //next button only enabled if length of username > 0
        mUsernameEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.length() > 0) {
                    ((StartupActivity)getActivity()).setEnableNextButton(true);
                }
                else{
                    ((StartupActivity)getActivity()).setEnableNextButton(false);
                }

                //make username all lowercase
                if (!editable.toString().equals(editable.toString().toLowerCase()))
                {
                    mUsernameEditText.setText(editable.toString().toLowerCase());
                    mUsernameEditText.setSelection(mUsernameEditText.getText().toString().length());
                }
            }
        });

        return v;

    }


    public void updateUsername() {
        String newUsername = mUsernameEditText.getText().toString();
        if (mDefaultUsername.equals(newUsername)) {
            SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getActivity().getApplicationContext());
            sharedPref.edit().putString(FacebookLoginFragment.USERNAME_SHAREDPREFS,newUsername).commit();
            ((StartupActivity) getActivity()).finishStartup();
        } else {
            UpdateUsernameRequest request = new UpdateUsernameRequest();
            request.setUsername(newUsername);
            ((StartupActivity) getActivity()).showProgressSpinner();
            new UpdateUsernameTask(getActivity(), this).execute(request);
        }
    }


        private static class UpdateUsernameTask extends AsyncTask<UpdateUsernameRequest, Void, AsyncTaskResult<UpdateUsernameResponse>>
        {
        private WeakReference<Activity> activityReference;
        private WeakReference<ChooseUsernameStartupFragment> fragmentReference;

        UpdateUsernameTask(Activity activity, ChooseUsernameStartupFragment fragment)
        {
            fragmentReference = new WeakReference<>(fragment);
            activityReference = new WeakReference<>(activity);
        }


        @Override
        protected AsyncTaskResult<UpdateUsernameResponse> doInBackground(UpdateUsernameRequest... params) {



        // Create a LambdaInvokerFactory, to be used to instantiate the Lambda proxy
        LambdaInvokerFactory factory = new LambdaInvokerFactory(
                activityReference.get().getApplicationContext(),
                Regions.US_EAST_1,
                FacebookLoginFragment.getCredentialsProvider(activityReference.get()));
         final LambdaFunctionsInterface lambdaFunctionsInterface = factory.build(LambdaFunctionsInterface.class);

                try {
                    UpdateUsernameResponse response = lambdaFunctionsInterface.UpdateUsername(params[0]);
                    return new AsyncTaskResult<>(response);
                } catch (LambdaFunctionException lfe) {
                        Log.i("ERROR", lfe.getDetails());
                        Log.i("ERROR",lfe.getStackTrace().toString());
                        lfe.printStackTrace();
                    return new AsyncTaskResult<>(lfe);
                }
                catch (AmazonServiceException ase) {
                    // invalid credentials, incorrect AWS signature, etc
                    Log.i("ERROR", ase.getErrorMessage());

                    return new AsyncTaskResult<>(ase);
                }
                catch (AmazonClientException ace) {
                    // Network issue
                    Log.i("ERROR", ace.toString());

                    return new AsyncTaskResult<>(ace);
                }
            }

            @Override
            protected void onPostExecute(AsyncTaskResult<UpdateUsernameResponse> asyncResult) {
                // get a reference to the activity and fragment if it is still there
                ChooseUsernameStartupFragment fragment = fragmentReference.get();
                Activity activity = activityReference.get();
                if (fragment == null || fragment.isRemoving()) return;
                if (activity == null || activity.isFinishing()) return;

                ((StartupActivity) activity).hideProgressSpiner();
                UpdateUsernameResponse result = asyncResult.getResult();
                if (asyncResult.getError() != null) {
                    new HandleLambdaError().handleError(asyncResult.getError(), activity,null);
                    return;
                }

                //Username Updated
                if (result != null) {
                    if (result.getResult().equals(UpdateUsernameResponse.getResultUsernameUpdated())) {
                        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(activity.getApplicationContext());
                        sharedPref.edit().putString(FacebookLoginFragment.USERNAME_SHAREDPREFS, fragment.mUsernameEditText.getText().toString()).commit();

                        ((StartupActivity) activity).finishStartup();
                    } else if (result.getResult().equals(ChooseUsernameStartupFragment.usernamameNotValidErorrCode)) {

                        Toast.makeText(activity, R.string.username_wrong_characters_toast, Toast.LENGTH_SHORT).show();
                    } else if (result.getResult().equals(ChooseUsernameStartupFragment.usernamameNotValidErorrCode)) {
                        Toast.makeText(activity, R.string.username_too_long_toast, Toast.LENGTH_SHORT).show();
                    } else if (result.getResult().equals(ChooseUsernameStartupFragment.usernamameNotValidErorrCode)) {
                        Toast.makeText(activity, R.string.username_already_exists_toast, Toast.LENGTH_SHORT).show();
                    }
                }


            }

    }





}

File: src/main/java/com/liamfarrell/android/snapbattle/mvvm_ui/startup/ChooseUsernameStartupFragment.kt
package com.liamfarrell.android.snapbattle.mvvm_ui.startup

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.preference.PreferenceManager
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.findNavController
import com.liamfarrell.android.snapbattle.R
import com.liamfarrell.android.snapbattle.databinding.FragmentChooseUsernameStartupBinding
import com.liamfarrell.android.snapbattle.di.Injectable
import com.liamfarrell.android.snapbattle.model.aws_lambda_function_deserialization.aws_lambda_functions.request.UpdateUsernameRequest
import com.liamfarrell.android.snapbattle.viewmodels.startup.ChooseUsernameStartupViewModel
import kotlinx.android.synthetic.main.fragment_choose_username_startup.*
import java.lang.ClassCastException
import javax.inject.Inject





class ChooseUsernameStartupFragment : Fragment() , Injectable {

    companion object{
        val usernamameNotValidErorrCode = "USERNAME_NOT_VALID"
        val usernameTooLongErrorCode = "USERNAME_TOO_LONG"
        val usernameUsernameAlreadyExistsErrorCode = "USERNAME_ALREADY_EXISTS"
    }

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory


    private lateinit var startupActivity : SetupToolbarInterface
    private lateinit var viewModel: ChooseUsernameStartupViewModel
    private lateinit var defaultUsername : String

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is SetupToolbarInterface){
            startupActivity = context
            startupActivity.setTitle(resources.getStringArray(R.array.startup_activity_titles)[3])
        } else {
            throw ClassCastException() }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        defaultUsername = arguments?.getString("defaultUsername") ?: ""
        setHasOptionsMenu(true)
        (activity as StartupActivity).disableNextButton()
        val binding = FragmentChooseUsernameStartupBinding.inflate(inflater, container, false)
        binding.lifecycleOwner = viewLifecycleOwner
        binding.usernameEditText.setText(defaultUsername)
        binding.usernameEditText.addTextChangedListener(getUsernameOnChangedListener())
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(ChooseUsernameStartupViewModel::class.java)
        binding.viewModel = viewModel
        subscribeUi()
        return binding.root
    }

    override fun onOptionsItemSelected(item : MenuItem) : Boolean {
        val id = item.itemId
        if (id == R.id.action_next) {
            updateUsername()
        } else if (id == R.id.action_skip) {
            nextActivity()
        }
        return false
    }

    private fun subscribeUi() {
        viewModel.errorMessage.observe(viewLifecycleOwner, Observer {
            it?.let{ Toast.makeText(context, it, Toast.LENGTH_SHORT).show()}
        })
    }

    private fun nextActivity() {
        findNavController().navigate(R.id.action_chooseUsernameStartupFragment_to_mainActivity)
    }

    @SuppressLint("ApplySharedPref")
    fun updateUsername() {
        val newUsername = usernameEditText.getText().toString()
        if (defaultUsername == newUsername) {
            val sharedPref = PreferenceManager.getDefaultSharedPreferences(activity!!.applicationContext)
            sharedPref.edit().putString(LoggedInFragment.USERNAME_SHAREDPREFS, newUsername).commit()
            nextActivity()
        } else {
            val request = UpdateUsernameRequest()
            request.username = newUsername
            viewModel.updateUsername(newUsername, ::nextActivity)
        }
    }


    private fun getUsernameOnChangedListener() = object : TextWatcher{
        override fun afterTextChanged(s: Editable?) {
            s?.let {
                if (s.isNotEmpty()) {
                    (activity as StartupActivity).enableNextButton()
                } else {
                    (activity as StartupActivity).disableNextButton()
                }
                //make username all lowercase
                if (s.toString() != s.toString().toLowerCase()) {
                    usernameEditText.setText(s.toString().toLowerCase())
                    usernameEditText.setSelection(usernameEditText.getText().toString().length)
                }
            }
        }
        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

    }



}